lib/internal/cluster/master.js

```diff
@@ -0,0 +1,383 @@
+    }
+  }
+
+  return fork(cluster.settings.exec, cluster.settings.args, {
+    env: workerEnv,
+    silent: cluster.settings.silent,
+    execArgv: execArgv,
+    stdio: cluster.settings.stdio,
+    gid: cluster.settings.gid,
+    uid: cluster.settings.uid
+  });
+}
+
+
+function removeWorker(worker) {
+  assert(worker);
+  delete cluster.workers[worker.id];
+
+  if (Object.keys(cluster.workers).length === 0) {
+    assert(Object.keys(handles).length === 0, 'Resource leak detected.');
+    intercom.emit('disconnect');
+  }
+}
+
+
+function removeHandlesForWorker(worker) {
+  assert(worker);
+
+  for (var key in handles) {
+    const handle = handles[key];
+
+    if (handle.remove(worker))
+      delete handles[key];
+  }
+}
+
+
+cluster.fork = function(env) {
+  cluster.setupMaster();
+  const id = ++ids;
+  const workerProcess = createWorkerProcess(id, env);
+  const worker = new Worker({
+    id: id,
+    process: workerProcess
+  });
+
+  worker.on('message', function(message, handle) {
+    cluster.emit('message', this, message, handle);
+  });
+
+  worker.process.once('exit', (exitCode, signalCode) => {
+    /*
+     * Remove the worker from the workers list only
+     * if it has disconnected, otherwise we might
+     * still want to access it.
+     */
+    if (!worker.isConnected()) {
+      removeHandlesForWorker(worker);
+      removeWorker(worker);
+    }
+
+    worker.exitedAfterDisconnect = !!worker.exitedAfterDisconnect;
+    worker.state = 'dead';
+    worker.emit('exit', exitCode, signalCode);
+    cluster.emit('exit', worker, exitCode, signalCode);
+  });
+
+  worker.process.once('disconnect', () => {
+    /*
+     * Now is a good time to remove the handles
+     * associated with this worker because it is
+     * not connected to the master anymore.
+     */
+    removeHandlesForWorker(worker);
+
+    /*
+     * Remove the worker from the workers list only
+     * if its process has exited. Otherwise, we might
+     * still want to access it.
+     */
+    if (worker.isDead())
+      removeWorker(worker);
+
+    worker.exitedAfterDisconnect = !!worker.exitedAfterDisconnect;
+    worker.state = 'disconnected';
+    worker.emit('disconnect');
+    cluster.emit('disconnect', worker);
+  });
+
+  worker.process.on('internalMessage', internal(worker, onmessage));
+  process.nextTick(emitForkNT, worker);
+  cluster.workers[worker.id] = worker;
+  return worker;
+};
+
+
+function emitForkNT(worker) {
+  cluster.emit('fork', worker);
+}
+
+
+cluster.disconnect = function(cb) {
+  const workers = Object.keys(cluster.workers);
+
+  if (workers.length === 0) {
+    process.nextTick(() => intercom.emit('disconnect'));
+  } else {
+    for (var key in workers) {
+      key = workers[key];
+
+      if (cluster.workers[key].isConnected())
+        cluster.workers[key].disconnect();
+    }
+  }
+
+  if (typeof cb === 'function')
+    intercom.once('disconnect', cb);
+};
+
+
+function onmessage(message, handle) {
```

Comment:
Nit: While we're refactoring all this code, maybe remove `handle` as an argument here? It's not used in the function.